### PR TITLE
Soletta meets JS v2

### PR DESCRIPTION
**Pull request by ricardotk**
DIfferences from v1:
    - start raising javascript exceptions in error cases
    - start comparing packet type pointers in `send_packet()`
    - remove useless `| 0` in `duk_def_prop()` flags
    - create auxiliary functions to process functions
    - use `strndupa()` instead of `sol_arena_strdup_slice()` in `create_js_type()`
    - small changes on samples

This series introduce Javascript node using Duktape[1] as third party library,
allowing the usage of Javascript language to create new and customizable node types.

The aim of this series is to introduce an initial support for Javascript nodes, there is
other features that we'll add later on.

A JS node type is specified with one object containing each input and output port
declarations (name and type) and its callback functions that will be trigged on the
occurrence of certain events like input/output ports processes, open/close
processes, so forth and so on.

### CODE REVIEW

File: src/lib/flow/sol-flow-js.c [hunk 1]
```diff
@@ -0,0 +1,1328 @@
...
+    duk_put_prop_string(mdata->duk_ctx, -2, "sendPacket");
+
+    duk_push_c_function(mdata->duk_ctx, send_error_packet, 2);
+    duk_put_prop_string(mdata->duk_ctx, -2, "sendErrorPacket");
+
+    /* From this point node JS object is always in the top of the stack. */
+    duk_get_prop_string(mdata->duk_ctx, -1, "node");
+
+    if (!setup_ports_methods(mdata->duk_ctx, type->ports_in.len, type->ports_out.len)) {
+        SOL_ERR("Failed to bla TODO: %s", duk_safe_to_string(mdata->duk_ctx, -1));
```
Comment (lpereira) [inline src/lib/flow/sol-flow-js.c]: :)

Comment (ricardotk) [inline src/lib/flow/sol-flow-js.c]: It was a test and you're approved! congratz (: :confetti_ball:

**Comment by ricardotk** [comment]
sparsecheckout seems to be the best solution among the others, thanks @bdilly and @dorileo

**Comment by ricardotk** [comment]
I changed to sparse checkout and added a MIT file about the license (required by legal), does anyone have more comments? I'll push it today if no other commentary shows up.

**Comment by ricardotk** [comment]
v3 PR #346

**Pull request closed without merging**